test: folders spec [WPB-19951]

JacquelineLehner opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-19951" title="WPB-19951" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />WPB-19951</a>  [Web/QA] Write the Folders regression tests in Playwright
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->


# Pull Request

## Summary
Adds the regression tests for creating custom folders and moving conversations into custom folders

---

## Security Checklist (required)

- [x] **External inputs are validated & sanitized** on client and/or server where applicable.
- [x] **API responses are validated**; unexpected shapes are handled safely (fallbacks or errors).
- [x] **No unsafe HTML is rendered**; if unavoidable, sanitization is applied **and** documented where it happens.
- [x] **Injection risks (XSS/SQL/command) are prevented** via safe APIs and/or escaping.

## Checklist

- [x] mentions the JIRA issue in the PR name ([WPB-19951])
- [x] PR has been self reviewed by the author;
- [x] Hard-to-understand areas of the code have been commented;
- [ ] If it is a core feature, unit tests have been added;


## Standards Acknowledgement (required)

- [x] I have read and this PR **upholds** our [Coding Standards](https://github.com/wireapp/wire-webapp/tree/docs/coding-standards.md) and [Tech Radar Choices](https://github.com/wireapp/wire-webapp/tree/docs/tech-radar.md).


[WPB-19951]: https://wearezeta.atlassian.net/browse/WPB-19951

JacquelineLehner committed:
test(folders): implemented folder tests according to TCF-1351 in Testiny

JacquelineLehner committed:
test(folders): add regression tests for custom folders

This commit introduces a new test suite `folders.spec.ts` to verify
the functionality of custom folders.

Implemented test cases include:
- Moving 1:1 and group conversations to new or existing folders
  (@TC-545, @TC-546, @TC-547, @TC-548).
- Validating that folders cannot be created without a name (@TC-560).
- Verifying that deleted folders leave no traces (@TC-568).

Additionally, a new `OptionModal` page object was created to handle
interactions with the folder creation dialog.

Note: @TC-553 is currently implemented but skipped due to the known
bug [WPB-22266].

Refs: WPB-19951

markbrockhoff commented on test/e2e_tests/pageManager/webapp/pages/conversationList.page.ts:
```diff
@@ -58,6 +61,11 @@ export class ConversationListPage {
+    this.moveToMenu = page.locator('ul.ctx-menu', {
+      has: page.locator(selectById('btn-create-new-folder')),
+    });
+    this.createNewFolderButton = this.moveToMenu.locator(selectById('btn-create-new-folder'));
```
Please try to use accessible locators where possible over (test-)ids. For example in this case:

```suggestion
    this.moveConversationButton = page.getByRole('menu').getByRole('button', {name: 'Move to'});
    this.moveToMenu = page.getByRole('menu');
    this.createNewFolderButton = this.moveToMenu.getByRole('button', {name: 'Create new folder'});
```

JacquelineLehner committed:
test(folders): overhaul testcases to fix code duplication

This commit overhauls the new test suite `folders.spec.ts` (to verify
the functionality of custom folders) to reduce code duplication.

Implemented test cases include:
- Moving 1:1 and group conversations to new or existing folders
  (@TC-545, @TC-546, @TC-547, @TC-548).
- Validating that folders cannot be created without a name (@TC-560).
- Verifying that deleted folders leave no traces (@TC-568).

Additionally, a new `OptionModal` page object was created to handle
interactions with the folder creation dialog.

Note: @TC-553 is currently implemented but skipped due to the known
bug [WPB-22266]. Code Duplication cannot be reduced any further.

Refs: WPB-19951

JacquelineLehner commented on test/e2e_tests/specs/Folders/folders.spec.ts:
```diff
@@ -0,0 +1,242 @@
+
+      // Preconditions: Create a custom folder with group conversation
+      await createCustomFolder(userAPageManagerInstance, userAModals.optionModal(), conversationName, customFolderName);
+      await userAComponents.conversationSidebar().allConverationsButton.click();
```
button was named with a typo in the first place

JacquelineLehner committed:
test(folders): overhaul testcases to fix PR comments

This commit overhauls the new test suite `folders.spec.ts` (to verify
the functionality of custom folders) according to PR comments from Copilot.

Implemented test cases include:
- Moving 1:1 and group conversations to new or existing folders
  (@TC-545, @TC-546, @TC-547, @TC-548).
- Validating that folders cannot be created without a name (@TC-560).
- Verifying that deleted folders leave no traces (@TC-568).

Additionally, a new `OptionModal` page object was created to handle
interactions with the folder creation dialog.

Note: @TC-553 is currently implemented but skipped due to the known
bug [WPB-22266]. Code Duplication cannot be reduced any further.

Refs: WPB-19951

markbrockhoff commented on test/e2e_tests/specs/Folders/folders.spec.ts:
```diff
@@ -0,0 +1,223 @@
+ */
+async function createCustomFolder(
+  pageManager: PageManager,
+  optionModal: OptionModal,
```
Isn't the option modal already included in `pageManager.webapp.modals`? I think you can just use it and safe the extra parameter :)

markbrockhoff commented on test/e2e_tests/specs/Folders/folders.spec.ts:
```diff
@@ -0,0 +1,223 @@
+    // Step 2: User A moves 1:1 conversation with User B into new custom folder
+    await createCustomFolder(userAPageManagerInstance, userAModals.optionModal(), userB.fullName, customFolderName);
+    // Step 3: 1:1 conversation with User B is in the custom folder
+    const actualTitle = userAPages.conversationList().page.locator('[data-uie-name="conversation-list-header-title"]');
```
You're using this locator 5 times within the test, I think it's ok to move it into the conversationList POM ;)

markbrockhoff commented on test/e2e_tests/specs/Folders/folders.spec.ts:
```diff
@@ -0,0 +1,223 @@
+  let userA: User;
+  let userB: User;
+  let userAPageManagerInstance: PageManager;
+  let userAPageManager: PageManager['webapp'];
```
Suggestion: Maybe storing the PageManagerInstance is enough.
You created two variables, one called PageManager but it's type is not the PageManager but it's subtype for the webapp. The other one is called "...Instance" just to not collide with the other variable name.

IMHO, it might be less confusing to rename `userAPageManagerInstance` to `userAPageManager` removing the current `userAPageManager` variable. You can just use `userAPageManager.webapp` where it was previously used instead.

markbrockhoff commented on test/e2e_tests/specs/Folders/folders.spec.ts:
```diff
@@ -0,0 +1,223 @@
+    await userAPages.conversationList().getRemoveConversationFromFolderButton(customFolderName).click();
+
+    // Step 2: User A tries to move conversation with User B back into the folder, custom folder name should no longer be visible in the folder menu
+    await userAPages.conversationList().openConversation(userB.fullName);
```
Valid finding

markbrockhoff replied:
Are you sure this was resolved? I fear the `openConversation()` might close the context menu which would make the assertion always pass. Could you maybe double check? :)

JacquelineLehner replied:
Copilot was right

markbrockhoff commented on test/e2e_tests/specs/Folders/folders.spec.ts:
```diff
@@ -0,0 +1,223 @@
+        .page.locator('[data-uie-name="conversation-list-header-title"]');
+
+      await expect(actualTitle).toHaveText('All Conversations');
+      // Step 3: Custom Folder Name is no longer visible in the Move-To-Menu
```
Also valid

JacquelineLehner committed:
test(folders): overhaul testcases according to PR comments

This commit overhauls the new test suite `folders.spec.ts` (to verify
the functionality of custom folders) according to PR comments.

Implemented test cases include:
- Moving 1:1 and group conversations to new or existing folders
  (@TC-545, @TC-546, @TC-547, @TC-548).
- Validating that folders cannot be created without a name (@TC-560).
- Verifying that deleted folders leave no traces (@TC-568).

Additionally, a new `OptionModal` page object was created to handle
interactions with the folder creation dialog.

Note: @TC-553 is currently implemented but skipped due to the known
bug [WPB-22266]. Code Duplication cannot be reduced any further.

Refs: WPB-19951

markbrockhoff commented on test/e2e_tests/specs/Folders/folders.spec.ts:
```diff
@@ -21,31 +21,25 @@ import {User} from 'test/e2e_tests/data/user';
-) {
+async function createCustomFolder(pageManager: PageManager, conversationName: string, folderName: string) {
   const pages = pageManager.webapp.pages;
+  const modals = pageManager.webapp.modals;
```
Nitpick:
```suggestion
  const {pages, modals} = pageManager.webapp;
```

JacquelineLehner replied:
changed

JacquelineLehner committed:
test(folders): overhaul testcases according to Review comments

This commit overhauls the new test suite `folders.spec.ts` (to verify
the functionality of custom folders) according to Review comments.

Implemented test cases include:
- Moving 1:1 and group conversations to new or existing folders
  (@TC-545, @TC-546, @TC-547, @TC-548).
- Validating that folders cannot be created without a name (@TC-560).
- Verifying that deleted folders leave no traces (@TC-568).

Additionally, a new `OptionModal` page object was created to handle
interactions with the folder creation dialog.

Note: @TC-553 is currently implemented but skipped due to the known
bug [WPB-22266]. Code Duplication cannot be reduced any further.

Refs: WPB-19951
